result/result.py

```diff
@@ -95,23 +98,23 @@ def value(self) -> Union[E, T]:
 
 
 @overload
-def Ok() -> Result[E, bool]:
+def Ok() -> Result[bool, Any]:
     pass
 
 
 @overload
-def Ok(value: T) -> Result[E, T]:
+def Ok(value: T) -> Result[T, Any]:
```

Comment:
This failed to work with latest mypy version when putting `E` in the return type. That makes sense, because mypy would not know what type E should be, and there is no class context like for `Result.Ok` where mypy asks for a type annotation.

The problem now is that `result.Result.Ok(...)` must be type annotated as `Result[T, E]` while `result.Ok(...)` is implicitly the type `Result[T, Any]` without mypy requesting a type annotation.

My suggestion would be to deprecate the helper functions, unless we find a way to properly annotate them. After all, from the Zen of Python:

> There should be one-- and preferably only one --obvious way to do it.

What do you think, @tyehle?

CC @snorfalorpagus

Reply:
Yes, with mypy 0.761, when doing this:

    res1 = result.Ok()
    res2 = result.Result.Ok()

You get this:

    result/typetests.py:15:8: error: Need type comment for 'res2'
    result/typetests.py:16:8: error: Need type comment for 'res3'

This makes sense, since the free-standing function is annotated with a generic type parameter that has no type bounds:

    @overload
    def Ok() -> Result[bool, E]:
        pass

Mypy cannot know what `E` is, so the user needs to add type annotations.

However, after thinking more about it, removing the helper functions won't help at all. On the contrary, returning `Result[T, E]` is wrong if we can't know what type `E` is. So `Result[T, Any]` *is* the correct type, which can later be refined e.g. when the result is mapped with `map_err`.